File: java/leetcode/editor/cn/RangeSumQueryMutable.java
package leetcode.editor.cn;

//[307] 区域和检索 - 数组可修改
/* ["NumArray","sumRange","update","sumRange"],[[[1,3,5]],[0,2],[1,2],[0,2]] */
public class RangeSumQueryMutable {
    public static void main(String[] args) {
    }

    //leetcode submit region begin(Prohibit modification and deletion)
    class NumArray {
        int[] tree;
        int[] nums;
        int n;

        int lowBit(int x) {
            return x & -x;
        }

        int query(int x) {
            int ans = 0;
            for (int i = x; i > 0; i -= lowBit(i)) {
                ans += tree[i];
            }
            return ans;
        }

        void add(int x, int u) {
            for (int i = x; i <= n; i += lowBit(i)){
                tree[i] += u;
            }
        }

        public NumArray(int[] nums) {
            this.nums = nums;
            n = this.nums.length;
            tree = new int[n + 1];
            for (int i = 0; i < n; i++){
                add(i + 1, this.nums[i]);
            }

        }

        public void update(int i, int val) {
            add(i + 1, val - nums[i]);
            nums[i] = val;
        }

        public int sumRange(int l, int r) {
            return query(r + 1) - query(l);
        }
    }
/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * obj.update(index,val);
 * int param_2 = obj.sumRange(left,right);
 */
//leetcode submit region end(Prohibit modification and deletion)

}

File: src/main/java/leetcode/editor/cn/RangeSumQueryMutable.kt
package leetcode.editor.cn

import org.hamcrest.MatcherAssert.*
import org.hamcrest.Matchers.*
import util.to1DIntArray

// 区域和检索 - 数组可修改
// 307
/* ["NumArray","sumRange","update","sumRange"],[[[1,3,5]],[0,2],[1,2],[0,2]] */

fun main() {
    val test1 = RangeSumQueryMutable.NumArray("[1,3,5]".to1DIntArray())
    assertThat(test1.sumRange(0, 2), equalTo(9))
    test1.update(1, 2)
    assertThat(test1.sumRange(0, 2), equalTo(8))

    val test2 = RangeSumQueryMutable.NumArray("[7,2,7,2,0]".to1DIntArray())
    test2.update(4, 6)      // 7,2,7,2,6
    test2.update(0, 2)      // 2,2,7,2,6
    test2.update(0, 9)      // 9,2,7,2,6
    assertThat(test2.sumRange(4, 4), equalTo(6))
    test2.update(3, 8)      // 9,2,7,8,6
    assertThat(test2.sumRange(0, 4), equalTo(32))
    test2.update(4, 1)      // 9,2,7,8,1
    assertThat(test2.sumRange(0, 3), equalTo(26))
    assertThat(test2.sumRange(0, 4), equalTo(27))
    test2.update(0, 4)      // 4,2,7,8,1
    assertThat(test2.sumRange(0, 4), equalTo(22))

}

class RangeSumQueryMutable {
    //leetcode submit region begin(Prohibit modification and deletion)
    class NumArray(nums: IntArray) {
        val BIT = IntArray(nums.size + 1)

        init {
            nums.withIndex().forEach { BIT[it.index + 1] = it.value }
            for (i in 1..BIT.lastIndex) {
                val j = i + (i and -i)  // bit[j]管理了该层的和
                if (j < BIT.size) BIT[j] += BIT[i]
            }
        }

        fun update(i: Int, `val`: Int) {
            var idx = i + 1
            val delta = `val` - sumRange(i, i)
            while (idx < BIT.size) {
                BIT[idx] += delta
                idx += (idx and -idx)
            }
        }

        fun prefixSum(`val`: Int): Int {
            var idx = `val` + 1
            var res = 0
            while (idx > 0) {
                res += BIT[idx]
                idx -= idx and (-idx)
            }
            return res
        }

        fun sumRange(i: Int, j: Int): Int {
            return prefixSum(j) - prefixSum(i - 1)
        }

    }

//leetcode submit region end(Prohibit modification and deletion)

}